### PR TITLE
Fix incorrect tab focus on close when preceding tab is disabled (#15186)

### PR DESCRIPTION
**Pull request by Shinu-Cherian**
Summary
This PR fixes an issue where closing a tab could incorrectly focus a disabled tab and break keyboard navigation.

Root Cause
In `tabs.tsx`, the `close_tab` function computed the new active tab index using a naive `clamp` operation. This logic failed to check whether the newly computed active tab was `disabled`. Consequently, closing a tab often incorrectly shifted focus to an adjacent disabled tab, which short-circuited the `on_key` event handler and completely broke keyboard navigation across tabs.

 Solution
Updated the `close_tab` logic to safely calculate the new active tab:
- If a tab to the left of the active tab is closed, the active tab's index is explicitly decremented to maintain the same visual active tab.
- If the currently active tab is closed, the function now iterates through the remaining tabs to find the closest **non-disabled** tab, preferring tabs to the right, and then falling back to the left. 
- This guarantees that keyboard focus and navigation are properly maintained when deleting tabs.

Testing Performed
- Reproduced issue #15186 locally with a disabled tab and verified that closing the active tab now elegantly shifts focus to a non-disabled tab.
- Rebuilt BokehJS and verified that TypeScript compilation passes successfully.

### CODE REVIEW

**Comment by bryevdv** [comment]
Thanks for the PR @Shinu-Cherian What is your bandwidth for additional work on this issue? I think the current PR would be improved by adding some tests, which we can help with suggestions, pointers if necessary.

File: bokehjs/src/lib/models/layouts/tabs.tsx [hunk 1]
```diff
@@ -140,7 +140,33 @@ export class TabsView extends LayoutDOMView {
...
+                new_active = k
+                found = true
+                break
+              }
+            }
+          }
+          if (!found) {
+            new_active = 0
+          }
+        }
```
Comment (bryevdv) [inline bokehjs/src/lib/models/layouts/tabs.tsx]: @Shinu-Cherian This is OK but we do try avoid `let` by assigning to const values from IIFEs i.e. I think the above code code could be made to look something like this, if you were interested to clean it up a little more:
```typescript
  const new_active = (() => {
    if (new_tabs.length == 0) {
      return -1
    } else if (j < active) {
      return active - 1
    } else if (j == active) {
      return find_activable(j, new_tabs)
    } else {
      return active
    }
  })()
```

Comment (Shinu-Cherian) [reply bokehjs/src/lib/models/layouts/tabs.tsx]: Thank you for the review and the suggestion! I definitely have the bandwidth to work on this.

I will refactor the close_tab function to use the IIFE and const as you suggested. I'll also add unit tests to cover this behavior and ensure the regression doesn't happen again.

I will push the updates to this branch shortly!

File: bokehjs/src/lib/models/layouts/tabs.tsx [hunk 2]
```diff
@@ -140,7 +140,29 @@ export class TabsView extends LayoutDOMView {
...
+            for (let k = j; k < new_tabs.length; k++) {
+              if (!new_tabs[k].disabled) {
+                return k
+              }
+            }
+            for (let k = j - 1; k >= 0; k--) {
+              if (!new_tabs[k].disabled) {
+                return k
+              }
+            }
```
Comment (bryevdv) [inline bokehjs/src/lib/models/layouts/tabs.tsx]: @Shinu-Cherian instead of these loops, I think this `j == active` branch can use this existing `find_activable` function?

https://github.com/bokeh/bokeh/blob/784d225625feb5b826929e357848de97f73fad6d/bokehjs/src/lib/models/layouts/tabs.tsx#L201

File: bokehjs/src/lib/models/layouts/tabs.tsx [hunk 3]
```diff
@@ -140,7 +140,29 @@ export class TabsView extends LayoutDOMView {
...
       const close_tab = (j: number = i) => {
         const new_tabs = remove_at(tabs, j)
-        const new_active = clamp(active, 0, new_tabs.length - 1)
+
+        const new_active = (() => {
+          if (new_tabs.length == 0) {
+            return 0
+          } else if (j < active) {
+            return active - 1
+          } else if (j === active) {
```
Comment (bryevdv) [inline bokehjs/src/lib/models/layouts/tabs.tsx]:
```suggestion
          } else if (j == active) {
```
no need for `===` here

File: bokehjs/src/lib/models/layouts/tabs.tsx (hunk 2 above)
Comment (Shinu-Cherian) [reply bokehjs/src/lib/models/layouts/tabs.tsx]: Thanks for pointing that out! I completely missed the find_activable function. I will update the logic to use it and push the changes shortly

File: bokehjs/src/lib/models/layouts/tabs.tsx (hunk 3 above)
Comment (bryevdv) [reply bokehjs/src/lib/models/layouts/tabs.tsx]: @Shinu-Cherian can you make this small change here and anywhere elsewhere? I think everything else is looking pretty good, but we don't need `===` in these places and it is inconsistent with the rest of the codebase.

File: bokehjs/src/lib/models/layouts/tabs.tsx [hunk 4]
```diff
@@ -140,7 +168,21 @@ export class TabsView extends LayoutDOMView {
...
+            const dir = j === new_tabs.length ? -1 : 1
+            const start = dir === -1 ? j - 1 : j
+            return find_activable(new_tabs, start, dir, 0)
```
Comment (bryevdv) [inline bokehjs/src/lib/models/layouts/tabs.tsx]: Is this quite right? If `j == new_tabs.length` we can only search the "lower" direction, but in the other case I think we might need to search both directions. I.e. if all the "higher" tabs are disabled we would need to search the ones lower as well, no?

Comment (bryevdv) [reply bokehjs/src/lib/models/layouts/tabs.tsx]: As one idea, I think if you get rid of `fallback` and just return `null` in case nothing activable is found, then something like this would cover all cases:
```typescript
return find_activable(new_tabs, j, +1) ?? find_activable(new_tabs, j - 1, -1) ?? 0
```
The other users of `find_activable` will also need slight adjustment e.g. ` find_activable(...) || active` as well

Comment (Shinu-Cherian) [reply bokehjs/src/lib/models/layouts/tabs.tsx]: That's a very elegant suggestion! However, the existing find_activable function actually has wrap-around logic built into it (if dir=1, it searches i to n-1 and then wraps to search 0 to i-1).

Because it wraps around, find_activable(new_tabs, j, +1) will almost never return null. Instead of failing and falling back to the left tab (j-1), it will wrap around and select the first tab (0).

If we want to use your exact logic (find_activable(new_tabs, j, +1) ?? find_activable(new_tabs, j - 1, -1) ?? 0), we would need to remove the wrap-around loops from find_activable. But if we remove that, it will change the keyboard navigation behavior (e.g., ArrowRight on the last tab will no longer wrap to the first tab).

How would you like to proceed?

Keep the wrap-around in find_activable and use a slightly customized condition in close_tab (similar to what I currently have).
Remove the wrap-around from find_activable so we can cleanly use the ?? fallback logic you suggested, assuming it's okay if keyboard navigation stops wrapping.
Also, I will definitely change the === to == as you pointed out!

Comment (bryevdv) [reply bokehjs/src/lib/models/layouts/tabs.tsx]: @Shinu-Cherian You're right I overlooked that `find_activable` already does consider both directions (i.e. that `dir` only specifies which direction to _start_ looking) In that case I think the I'd only suggest simplifying and removing the `fallback` parameter and just always returning null in that case. Then the code here would just be `return find_activable(...) || 0` and the other places would be `toggle_tab(find_activable(...) || active)` that should not change any behaviour unless I am mistaken.

Comment (Shinu-Cherian) [reply bokehjs/src/lib/models/layouts/tabs.tsx]: Makes total sense! I've removed the fallback parameter, used the nullish coalescing operator (??), and changed the === to == as you suggested. The updated changes have been pushed!

File: bokehjs/test/unit/models/layouts/tabs.ts [hunk 5]
```diff
@@ -39,4 +39,24 @@ describe("Tabs", () => {
...
+
+    const view = await build_view(tabs)
+    view.render_to(document.body)
+
+    const close_btns = view.shadow_el.querySelectorAll(".bk-close")
+    const close_btn = close_btns[1] as HTMLElement // The second tab is active, index 1
+    close_btn.click()
+
+    expect(tabs.active).to.be.equal(1) // was index 2, now shifted to index 1
+    expect(tabs.tabs.length).to.be.equal(2)
```
Comment (bryevdv) [inline bokehjs/test/unit/models/layouts/tabs.ts:60]: This is a good test, but it would be nice to add a few more to cover the other different cases --- zero tabs left, zero activable tabs left, cases with an activable tab that is "lower" or "higher" than original active index, etc.

**Review by bryevdv — COMMENTED** [review]
@Shinu-Cherian I only have one final comment, if you have bandwith, which is just to flesh out the tests to include more cases. Otherwise the implementation looks good to go

File: bokehjs/test/unit/models/layouts/tabs.ts (hunk 5 above)
Comment (Shinu-Cherian) [reply bokehjs/test/unit/models/layouts/tabs.ts:60]: Absolutely! I will add the additional test cases you mentioned to cover all those edge cases (zero tabs, zero activable tabs, higher/lower tabs) and push the updates shortly.

**Review by bryevdv — APPROVED** [review]

**Comment by bryevdv** [comment]
Thanks @Shinu-Cherian There are some unrelated test failures. I've gone ahead and approved this, but I would like to try and sort those out in the next day or two and see this green before merging. We will get it merged this week regardless.

**Comment by Shinu-Cherian** [comment]
> Thanks @Shinu-Cherian There are some unrelated test failures. I've gone ahead and approved this, but I would like to try and sort those out in the next day or two and see this green before merging. We will get it merged this week regardless.


Awesome, thank you so much for the approval and for taking the time to review my PR! Take your time sorting out the CI failures. I'm really excited to see this merged soon. Thanks again!
